mirror/mirror.go

```diff
@@ -0,0 +1,371 @@
+	gn := gsnet.NewFromLibp2pHost(host)
+	gx := gsimpl.New(ctx, gn, ls)
+	dn := dtnetwork.NewFromLibp2pHost(host)
+	tp := gstransport.NewTransport(host.ID(), gx)
+	dm, err := datatransfer.NewDataTransfer(ds, dn, tp)
+	if err != nil {
+		return nil, nil, err
+	}
+	dtReady := make(chan error)
+	dm.OnReady(func(e error) { dtReady <- e })
+	if err := dm.Start(ctx); err != nil {
+		return nil, nil, err
+	}
+	if err := <-dtReady; err != nil {
+		return nil, nil, err
+	}
+	return dm, gx, nil
+}
+
+func (m *Mirror) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	m.cancel = cancel
+	go func() {
+		for {
+			var t time.Time
+			select {
+			case t = <-m.ticker.C:
+			case <-ctx.Done():
+				return
+			}
+			log := log.With("time", t)
+
+			mc, err := m.getLatestOriginalAdCid(ctx)
+
+			if err != nil {
+				log.Errorw("failed to get the latest mirrored cid", "err", err)
+				continue
+			}
+			log = log.With("latestMirroredCid", mc)
+
+			var syncedAdCids []cid.Cid
+			var sel ipld.Node
+			if cid.Undef.Equals(mc) {
+				sel = selectors.adsWithRecursionLimit(m.initAdRecurLimit)
+			} else {
+				sel = selectors.adsWithStopAt(selector.RecursionLimitNone(), cidlink.Link{Cid: mc})
+			}
+
+			_, err = m.sub.Sync(ctx, m.source.ID, cid.Undef, sel, m.source.Addrs[0],
+				legs.ScopedBlockHook(func(id peer.ID, c cid.Cid, actions legs.SegmentSyncActions) {
+					// TODO: set actions next segment link to ad previous id if it is present. For
+					//      now segmentation is disabled.
+					//       Here we could be encountering HAMT or Entry Chunk so picking the next
+					//       CID is not trivial; we probably should not use segmentation for HAMT
+					//       at all.
+
+					// Prepend to the list since the mirroring should start from the oldest ad first.
+					syncedAdCids = append([]cid.Cid{c}, syncedAdCids...)
+				}),
+				// Disable segmentation until the actions in hook are handled appropriately
+				legs.ScopedSegmentDepthLimit(-1),
+			)
+
+			if err != nil {
+				log.Errorw("Failed to sync source", "err", err)
+				continue
+			}
+
+			for _, adCid := range syncedAdCids {
+				err := m.mirror(ctx, adCid)
+				if err != nil {
+					log.Errorw("Failed to mirror ad", "cid", adCid, "err", err)
+					// TODO add an option on what to do if the mirroring of an ad failed?
+				}
+			}
+
+			syncedCount := len(syncedAdCids)
+			if syncedCount > 0 {
+				latestOriginal := syncedAdCids[syncedCount-1]
+				err = m.setLatestOriginalAdCid(ctx, latestOriginal)
+				if err != nil {
+					log.Errorw("Failed to store latest original ad cid", "cid", latestOriginal, "err", err)
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (m *Mirror) Shutdown() error {
+	if m.cancel != nil {
+		m.cancel()
+	}
+	return nil
+}
+
+func (m *Mirror) mirror(ctx context.Context, adCid cid.Cid) error {
+	ad, err := m.loadAd(ctx, adCid)
+	if err != nil {
+		return err
+	}
+	if err := ad.Validate(); err != nil {
```

Comment:
Is there any need to also check the signature on the ad? If a remote mirror is running, possibly mirroring specific providers, then that mirror may want to verify the ad signature.